### DISCUSSION TITLE
feedback of the shimming result

### DISCUSSION THREAD

**Discussion by gxu32** [shimming-toolbox/shimming-toolbox#519] [Q&A] [answered]
Dear experts:
Accroding to your kind advices, I have made a test on our machine. Our MRI can automatically generate the B0 fieldmap from the sequence. The commercial MRCodetool aslo use the same B0 fieldmap to calcutae the shimming coefficents, their result seems to be good. Below is the original B0 filedmap, the B0 fieldmap after using MRCodetool and 
the B0 fieldmap obtained by applying the shimming coefficents calculated by shimmingToolbox into the B0 fieldmap sequence on our MRI :
![1](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/4cdc57b1-67ea-4990-bca4-498722749e0f)

below is the shimming coefficents obtained from MRCodetool and shimmingToolbox：
![2](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/3783abe7-2d87-4ecd-9a0b-b3a3068ea425)

Although both MRCodetool and shimmingToolbox use the same original B0 filedmap for shimming coefficents calculation, it seems that the MRCodetool is more homogeneous at the region of back of head.
Below are the mask I used for shimmingToolbox and the mask created and used by MRCodetool:
![3](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/60f06623-7755-4682-9a53-58cf07e82ba6)

I checked the log file of MRCodetool, I found MRCodetool have a extra step called B0 unwrapping, I have no idea if it is this step that make the original B0 filedmap more accurate and hence make the difference between the MRCodetool and shimmingToolbox. Below is the part log file of MRCodetool:
![4](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/abc2f865-8d0f-4ef3-ab20-0e9b70aa767f)

I also want to get the B0 fieldmap from the initial phase map by myself, But from the json file of our phase image, I found it lack some critical infomation, Such as ΔTE, echo number. below is the json file of our phase image:
![5](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/008333ba-a52b-4041-b99e-289286124e32)

now I am tring to get these missing infomation directly from our sequence and add them to the json file to make the create fieldmap module of shimmingToolbox running.
 
Many thanks for your help!
best,
gaoqiang

**Comment by po09i** [comment]
Thanks @gxu32 for showing your results!

Shimming Toolbox expects an unwrapped field map as an input. If the field map provided has wraps, it could explain why the lower part of the head (where it is wrapped in the original field map) is not properly shimmed. If you want I can take a look at your data (DICOMs) if you send it to me (alexandre.dastous@polymtl.ca) and I can report if the information is available and if not, if I can update Shimming Toolbox to use the available information. 

Best,
Alex

**Reply by gxu32** [reply]
thank you Doctor Alex! @po09i right now, I only have the par/rec file, or nifti file obtained from dcm2niix conversion. and I'm going to do the scan to get the dicoms as soon as possible! Thanks again for your time!
best,
gaoqiang xu

**Comment by gxu32** [comment]
Hi Alex 
I'm really grateful to you for your help @po09i. And as you suggested, I think our raw B0 filedmap is wrapped which may be the reason for the poor B0 shimming result. I have sent the raw B0 filedmap (dicom files) to you. Thanks again! 
I'm also trying to use the raw phase map to get the fieldmap by myself. However, our sequence only generates one phase map, I presume it is the phase difference map of two echos. And after converting this phase map (par/rec) to nifti using dcm2niix, in the json file we only find one echo time and no echo space. So, we have to look for the echo space in the examcard of the sequence. Then we add the "EchoTime1" and "EchoTime2" tag (calculated by the echo time and the echo space) into the json file based on the JSON file from the spine example you provided. But when we are using the filedmap module of the shimmingtoolbox to get the filedmap, an error occurs as following:
INFO:shimmingtoolbox.load_nifti:Scaling the selected NIfTI 
… …
… … 
File /root/shimming-toolbox/shimming-toolbox/shimmingtoolbox/prepare_fieldmap.py, line 55, in prepare_fieldmap
raise ValueError("Values must range from -pi to pi.")
ValueError: Values must range from -pi to pi.
ERROR: 1

the phase map is int16 datatype, and if apply the "rescaledslope" and "rescaledintercept" in the header of the phase map nifti file, we are sure that the phase is in the range of [-pi,pi]. We have no idea why this error occur, could you give us some suggestions?

best,
gaoqiang

**Reply by gxu32** [reply]
Hi Alex
@po09i, I'm trying to send the dicom files to you using this address (alexandre.dastous@polymtl.ca), but  it seems this email address is invalid now, So I sent it to your another address (adastous023@gmail.com). Thanks!
best,
gaoqiang

**Reply by po09i** [reply]
I was able to look at your data (field map data) and confirmed that the wrapping issue should get fixed prior to shimming using Shimming Toolbox. Using the phase map(s) and unwrapping them should fix the issue using the field mapping pipeline. If you do not mind sending me your phase/mag data and I can take a look at the issue. I'm guessing your phase data might be slightly different than prior data and having this dataset could improve Shimming Toolbox's Philips support.

**Comment by gxu32** [comment]
Hi Alex
@po09i,I figured out what's going on with our phase map, our phase difference map is int16 at the range of [0,4094],this is very similar with the siemens field in your code of the load_nifti.py file. So I just change the tag "Manufacturer" in the json file from “Philips” to “siemens”. In addition, in your shimmingtoolbox. prepare_fieldmap.py, there is some code to make sure the range of the phase range is in [-math.pi,math.pi], but for our data, the minimum phase is -3.1416 which is a little smaller than -math.pi (-3.141592665), So the error occurs. So I change all 0 value in our phase map to be 1 to make it a little bigger, and we can get the fieldmap successfully. And we also do the B0 shimming based on this fieldmap, it seems the simulated result is good. We are going to test it on our 7T machine.
Thank you for you consistent help!
Best,
gaowqiang

**Reply by po09i** [reply]
That's great that you were able to get it working! I would love to implement a more permanent solution so that you and other Philips users do not have to manually do these manipulations. You can send me your data when you have a minute. 
Best,
Alex

**Reply by gxu32** [reply]
hi, Alex
I have sent it to you again this morning(<adastous023@gmail.com>), I don't know if you have receive it?
I think the simulated result is good , but the shimming coefficients seems to be alittle strange, I don't know if these coefficients can be applied to the MRI successfuly. I will try it tomorrow. the  coefficients  are: 22.502332, 26.849028, -247.935896, -13.848927, 3.720590, 3.343269, 0.395005, -0.471272,  they are seems much bigger. and the image is the simulated result: 
![fig_shimmed_vs_unshimmed](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/372ae6c7-c2be-4d85-bb23-f1bf73e9fa8b)
Thanks!

**Reply by po09i** [reply]
I believe the coefficients are now scaled for Siemens since the "Manufacturer" was now changed to Siemens.

**Comment by gxu32** [comment]
Hi Alex
Thanks, I will rescan the phantom agian in this week to get the phase/mag file in diocm format.
best,
gaoqiang

________________________________
发件人: Alex Dastous ***@***.***>
发送时间: 2024年3月18日 15:15
收件人: shimming-toolbox/shimming-toolbox ***@***.***>
抄送: Xu, Gaoqiang ***@***.***>; Mention ***@***.***>
主题: Re: [shimming-toolbox/shimming-toolbox] feedback of the shimming result (Discussion #519)

This Message Is From an External Sender
This message came from outside your organization.
Do not open any links or attachments unless you recognize the sender and were expecting the link or attachment.
Report Suspicious<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/CR-rLYS5ufxg2No!G9LdT0HCctgvIFNNXOEA-bQCFUntVFlv63CXVi9WuzfYdhyS5MiS3CIOAEoJjUs4d-9n67cRkZfuX0GNiKbQmSpkhusdC_3q6n-XAcst2RSQMCtXk809tJqGyJi8aEFxBHIfCLU8BXH-RCey9PsJ$>


I was able to look at your data (field map data) and confirmed that the wrapping issue should get fixed prior to shimming using Shimming Toolbox. Using the phase map(s) and unwrapping them should fix the issue using the field mapping pipeline. If you do not mind sending me your phase/mag data and I can take a look at the issue. I'm guessing your phase data might be slightly different than prior data and having this dataset could improve Philips support.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox/discussions/519*discussioncomment-8832228__;Iw!!CR-rLYS5ufxg2No!YutVFokNWIyUIr9Gi-0-xDfSFoFg5X-bErsS86ZTiyAVYlrG95sbn7waWv3n14eBk6qsSXNlvh1LF4Y7yvkzDOwmPryL$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFYAFPJNXUNT5K7PEDNWI3DYY44NRAVCNFSM6AAAAABENM4T4OVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DQMZSGIZDQ__;!!CR-rLYS5ufxg2No!YutVFokNWIyUIr9Gi-0-xDfSFoFg5X-bErsS86ZTiyAVYlrG95sbn7waWv3n14eBk6qsSXNlvh1LF4Y7yvkzDPkrI9_n$>.
You are receiving this because you were mentioned.

----------------------------------------------------------------------
Disclaimer:


The materials in this e-mail are private and may contain Protected Information. Please note that e-mail communication is not encrypted by default. You have the right to request further emails be encrypted by notifying the sender. Your continued use of e-mail constitutes your acknowledgment of these confidentiality and security limitations. If you are not the intended recipient, be advised that any unauthorized use, disclosure, copying, distribution, or the taking of any action in reliance on the contents of this information is strictly prohibited. If you have received this e-mail in error, please immediately notify the sender via telephone or return e-mail.

**Comment by gxu32** [comment]
Thanks! I got it, I think it is highly possible that our sequence dosen't fit the toolbox very well. It seems our sequence only have one echo and the phase images of this echo was used to calculate the filedmap, So I'm trying to make some adjust and report the result to you after the test! Thanks again!
________________________________
发件人: Alex Dastous ***@***.***>
发送时间: 2024年3月19日 11:40
收件人: shimming-toolbox/shimming-toolbox ***@***.***>
抄送: Xu, Gaoqiang ***@***.***>; Mention ***@***.***>
主题: Re: [shimming-toolbox/shimming-toolbox] feedback of the shimming result (Discussion #519)

This Message Is From an External Sender
This message came from outside your organization.
Do not open any links or attachments unless you recognize the sender and were expecting the link or attachment.
Report Suspicious<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/CR-rLYS5ufxg2No!G9LdT0HCctgvIFNNXOEA-bQCFUntVFlv63CXVi9WuzfYdhyS5MiS3CIOAEoJjUs4d-9n67cRkZfuX0GNiKbQmSpkhusd6_0KK_-XAcst2RQqskZuGLPohhQlQWZzFqnaZ9cP2lS8Vob_OicV_Sv3$>


I believe the coefficients are now scaled for Siemens since the "Manufacturer" was now changed to Siemens.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox/discussions/519*discussioncomment-8842544__;Iw!!CR-rLYS5ufxg2No!dvC0jUXj-yhxyUt1vsaF3_S6Yir6uu0wKN0ap279rcPIzvh_XaJPEB3BqX192iXKlQPInQ4UjNmj_6xyvNe_KwtYX-Yj$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFYAFPNKSFQBF2CUGMQ6VW3YZBL5XAVCNFSM6AAAAABENM4T4OVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DQNBSGU2DI__;!!CR-rLYS5ufxg2No!dvC0jUXj-yhxyUt1vsaF3_S6Yir6uu0wKN0ap279rcPIzvh_XaJPEB3BqX192iXKlQPInQ4UjNmj_6xyvNe_K-eT-ARh$>.
You are receiving this because you were mentioned.

----------------------------------------------------------------------
Disclaimer:


The materials in this e-mail are private and may contain Protected Information. Please note that e-mail communication is not encrypted by default. You have the right to request further emails be encrypted by notifying the sender. Your continued use of e-mail constitutes your acknowledgment of these confidentiality and security limitations. If you are not the intended recipient, be advised that any unauthorized use, disclosure, copying, distribution, or the taking of any action in reliance on the contents of this information is strictly prohibited. If you have received this e-mail in error, please immediately notify the sender via telephone or return e-mail.

**Comment by po09i** [comment, marked as answer]
Hi @gxu32,
The issue is that the input field map needs to be unwrapped when fed to the b0 shimming pipeline. Since your field map is already in Hz. I think that the best way to move forward is to create a new CLI (`st_unwrap`) that unwraps arbitrary data. I added some quality of life options for common data such as field maps (using `--unit` and `--dte`). You can therefore use your field map that is wrapped (in Hz) and provide the right options to unwrap it. The feature is still a work in progress (WIP) so to test it, you need additional manipulations that I provided below. You can look at the available options by using `st_unwrap -h`

Here is a way to test the branch:

```
cd <working_directory>
git clone https://github.com/shimming-toolbox/shimming-toolbox.git
cd shimming-toolbox
git checkout ad/philips-fmap
make install

st_unwrap -i </path/to/wrapped.nii.gz> --mag </path/to/input.nii.gz> -o </path/to/output.nii.gz> --unit hz --dte 0.001
```

Here are the results when I use it on your wrapped data:
<img src="https://github.com/shimming-toolbox/shimming-toolbox/assets/47249340/eb7c02da-924c-4d22-90e3-ad1d68be0279" width="400" height="400" />

Unwrapped:
<img src="https://github.com/shimming-toolbox/shimming-toolbox/assets/47249340/4c419927-1f6d-4e87-94db-2e5e9a2e547e" width="400" height="400" />

You should then be able to use the unwrapped field map to shim using `st_b0shim`. Let me know if that works or if you run into additional issues! 
Alex

**Reply by gxu32** [reply]
Hi Alex @po09i 
sorry for the delayed reply, with the implementation of your new function, I've done the B0 unwrapping step and tested the shim coefficients on our machine. The final results show a close similarity between the shimming toolbox and our MrCodetool. I'm truly grateful for your assistance! Below is a comparison of the B0 map after shimming with both tools.
![shimmed](https://github.com/shimming-toolbox/shimming-toolbox/assets/158335677/086664bb-f9c9-4c7b-b086-9d908dde2feb)
Next, I'm going to do B1 shimming on our 8 channels tranmit coils.
Thanks again!
best wishes
gaoqiang

**Reply by po09i** [reply]
Great to see similar results to other tools! As always, I'm happy to help for B1 shimming as well. 
Best,
Alex

**Comment by gxu32** [comment]
Hi Alex, Thank you so much for your time and help! I will test it again as soon as possible.

________________________________
发件人: Alex Dastous ***@***.***>
发送时间: 2024年3月21日 21:53
收件人: shimming-toolbox/shimming-toolbox ***@***.***>
抄送: Xu, Gaoqiang ***@***.***>; Mention ***@***.***>
主题: Re: [shimming-toolbox/shimming-toolbox] feedback of the shimming result (Discussion #519)

This Message Is From an External Sender
This message came from outside your organization.
Do not open any links or attachments unless you recognize the sender and were expecting the link or attachment.
Report Suspicious<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/CR-rLYS5ufxg2No!G9LdT0HCctgvIFNNXOEA-bQCFUntVFlv63CXVi9WuzfYdhyS5MiS3CIOAEoJjUs4d-9n67cRkZfuX0GNiKbQmSpkhusdi_0Kyx-XAcst2RT3aSy65yQTtGaYiDLcjPBQg_5XNrXdT3HWOHmCIFeN$>


Hi @gxu32<https://urldefense.com/v3/__https://github.com/gxu32__;!!CR-rLYS5ufxg2No!c9dkREV_5faMjamoSIhxqi7Y7OO782odszpMEIrH94-Nyfl1nwpeH7IfeGAgyPZQXLDAG34zrR_FUlX9nEY9c20q4-H9$>,
I took a look at your data and thought that the best way to move forward would be to create a new CLI (st_unwrap) that would unwrap arbitrary data. I added some quality of life options for common data such as field maps (using --unit and --dte). You can therefore use your field map that is wrapped (in Hz) and provide the right options to unwrap it. The feature is still a work in progress (WIP) so to test it, you need additional manipulations that I provided below. You can look at the available options by using st_unwrap -h

Here is a way to test the branch:

cd <working_directory>
git clone https://github.com/shimming-toolbox/shimming-toolbox.git<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox.git__;!!CR-rLYS5ufxg2No!c9dkREV_5faMjamoSIhxqi7Y7OO782odszpMEIrH94-Nyfl1nwpeH7IfeGAgyPZQXLDAG34zrR_FUlX9nEY9c5DHsnDH$>
cd shimming-toolbox
git checkout ad/philips-fmap
make install

st_unwrap -i </path/to/wrapped.nii.gz> --mag </path/to/input.nii.gz> -o </path/to/output.nii.gz> --unit hz --dte 0.001


Here are the results when I use it on your wrapped data:
Screenshot.2024-03-21.at.9.42.11.PM.png (view on web)<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox/assets/47249340/eb7c02da-924c-4d22-90e3-ad1d68be0279__;!!CR-rLYS5ufxg2No!c9dkREV_5faMjamoSIhxqi7Y7OO782odszpMEIrH94-Nyfl1nwpeH7IfeGAgyPZQXLDAG34zrR_FUlX9nEY9c18JujI-$>

Unwrapped:
Screenshot.2024-03-21.at.9.42.27.PM.png (view on web)<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox/assets/47249340/4c419927-1f6d-4e87-94db-2e5e9a2e547e__;!!CR-rLYS5ufxg2No!c9dkREV_5faMjamoSIhxqi7Y7OO782odszpMEIrH94-Nyfl1nwpeH7IfeGAgyPZQXLDAG34zrR_FUlX9nEY9cyskVHvE$>

You should then be able to use the unwrapped field map to shim using st_b0shim. Let me know if that works or if you run into additional issues!
Alex

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox/discussions/519*discussioncomment-8872533__;Iw!!CR-rLYS5ufxg2No!c9dkREV_5faMjamoSIhxqi7Y7OO782odszpMEIrH94-Nyfl1nwpeH7IfeGAgyPZQXLDAG34zrR_FUlX9nEY9c8CjbgVZ$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFYAFPOYCCI24KSUKSSZUULYZOFKFAVCNFSM6AAAAABENM4T4OVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DQNZSGUZTG__;!!CR-rLYS5ufxg2No!c9dkREV_5faMjamoSIhxqi7Y7OO782odszpMEIrH94-Nyfl1nwpeH7IfeGAgyPZQXLDAG34zrR_FUlX9nEY9c79-4gf5$>.
You are receiving this because you were mentioned.

----------------------------------------------------------------------
Disclaimer:


The materials in this e-mail are private and may contain Protected Information. Please note that e-mail communication is not encrypted by default. You have the right to request further emails be encrypted by notifying the sender. Your continued use of e-mail constitutes your acknowledgment of these confidentiality and security limitations. If you are not the intended recipient, be advised that any unauthorized use, disclosure, copying, distribution, or the taking of any action in reliance on the contents of this information is strictly prohibited. If you have received this e-mail in error, please immediately notify the sender via telephone or return e-mail.

**Comment by gxu32** [comment]
Many thanks to you for providing the excellent toolbox!
Best,
gaoqiang
________________________________
发件人: Alex Dastous ***@***.***>
发送时间: 2024年3月31日 13:43
收件人: shimming-toolbox/shimming-toolbox ***@***.***>
抄送: Xu, Gaoqiang ***@***.***>; Mention ***@***.***>
主题: Re: [shimming-toolbox/shimming-toolbox] feedback of the shimming result (Discussion #519)

This Message Is From an External Sender
This message came from outside your organization.
Do not open any links or attachments unless you recognize the sender and were expecting the link or attachment.
Report Suspicious<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/CR-rLYS5ufxg2No!G9LdT0HCctgvIFNNXOEA-bQCFUntVFlv63CXVi9WuzfYdhyS5MiS3CIOAEoJjUs4d-9n67cRkZfuX0GNiKbQmSpkhus9q32Ky9-XAcst2RTh6C8T7Zw4oliIm6Qz0mL7Bu_oUUhCegpw1F7fO6WU$>


Great to see similar results to other tools! As always, I'm happy to help for B1 shimming as well.
Best,
Alex

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/shimming-toolbox/shimming-toolbox/discussions/519*discussioncomment-8966135__;Iw!!CR-rLYS5ufxg2No!Z0BCo1jQn6AHlPVXynFciyaelAV9_RpAsFkpe-ufJSDiVjPsWtUa1OxmQx4qajFiiKXt3CZOy4AUC_Tv9xDk4gKrsGG9$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFYAFPMBQPBXCJVC5AXTN5TY3BDLDAVCNFSM6AAAAABENM4T4OVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DSNRWGEZTK__;!!CR-rLYS5ufxg2No!Z0BCo1jQn6AHlPVXynFciyaelAV9_RpAsFkpe-ufJSDiVjPsWtUa1OxmQx4qajFiiKXt3CZOy4AUC_Tv9xDk4pmFEAEw$>.
You are receiving this because you were mentioned.

----------------------------------------------------------------------
Disclaimer:


The materials in this e-mail are private and may contain Protected Information. Please note that e-mail communication is not encrypted by default. You have the right to request further emails be encrypted by notifying the sender. Your continued use of e-mail constitutes your acknowledgment of these confidentiality and security limitations. If you are not the intended recipient, be advised that any unauthorized use, disclosure, copying, distribution, or the taking of any action in reliance on the contents of this information is strictly prohibited. If you have received this e-mail in error, please immediately notify the sender via telephone or return e-mail.
